extensions/websockets-next/deployment/src/test/java/io/quarkus/websockets/next/test/telemetry/AbstractWebSocketsOnMessageTest.java

```diff
@@ -0,0 +1,349 @@
+
+    @TestHTTPResource("received-multi-dto-response-none")
+    URI multiDtoReceived_NoResponse_Uri;
+
+    @TestHTTPResource("received-multi-dto-response-single-dto")
+    URI multiDtoReceived_SingleDtoResponse_Uri;
+
+    @TestHTTPResource("received-multi-dto-response-multi-dto")
+    URI multiDtoReceived_MultiDtoResponse_Uri;
+
+    @TestHTTPResource("broadcast")
+    URI broadcast_Uri;
+
+    @Inject
+    Vertx vertx;
+
+    protected abstract boolean binaryMode();
+
+    protected abstract WebSocketConnector<?> bounceClientConnector();
+
+    protected abstract WebSocketConnector<?> multiClientConnector();
+
+    @Test
+    public void testServerEndpoint_SingleTextReceived_NoSent() {
+        // endpoint: void onMessage(String message)
+        var connection = Connection.of(singleTextReceived_NoResponse_Uri, false, 0, binaryMode(), "Ballad of a Prodigal Son");
+        sendAndAssertResponses(vertx, connection);
+        asserter.assertMetrics(0, 1, connection);
+    }
+
+    @Test
+    public void testServerEndpoint_SingleTextReceived_SingleTextSent() {
+        // endpoint: String onMessage(String message)
+        var connection = Connection.of(bounceUri, false, 1, binaryMode(), "Can't Find My Way Home");
+        sendAndAssertResponses(vertx, connection);
+        asserter.assertMetrics(0, 1, connection);
+    }
+
+    @Test
+    public void testServerEndpoint_SingleTextReceived_MultiTextSent() {
+        // endpoint: Multi<String> onMessage(String message)
+        var connection = Connection.of(singleTextReceived_MultiTextResponse_Uri, false, 2, binaryMode(),
+                "Always take a banana to a party");
+        sendAndAssertResponses(vertx, connection);
+        asserter.assertMetrics(0, 1, connection);
+    }
+
+    @Test
+    public void testServerEndpoint_MultiTextReceived_NoSent() {
+        // endpoint: void onMessage(Multi<String> message)
+        var connection = Connection.of(multiTextReceived_NoResponse_Uri, false, 0, binaryMode(), "When I go",
+                "don't cry for me",
+                "In my Father's arms I'll be", "The wounds this world left on my soul");
+        sendAndAssertResponses(vertx, connection);
+        asserter.assertMetrics(0, 4, connection);
+    }
+
+    @Test
+    public void testServerEndpoint_MultiTextReceived_SingleTextSent() {
+        // endpoint: String onMessage(Multi<String> message)
+        var connection = Connection.of(multiTextReceived_SingleTextResponse_Uri, false, 1, "Alpha Shallows", binaryMode(),
+                "Msg1", "Msg2", "Msg3", "Msg4");
+        sendAndAssertResponses(vertx, connection);
+        asserter.assertMetrics(0, 4, connection);
+    }
+
+    @Test
+    public void testServerEndpoint_MultiTextReceived_MultiTextSent() {
+        // endpoint: Multi<String> onMessage(Multi<String> message)
+        var connection = Connection.of(multiTextReceived_MultiTextResponse_Uri, false, 2, binaryMode(), "Msg1", "Msg2", "Msg3");
```

Comment:
I don't understand the effects of the multiplier on the test assertions... 
If the inputs are different the assertions should also have changed, no?

Reply:
when endpoint returns void, it is zero response
when endpoint returns one response, it is 1
when endpoint returns two results, it is 2
and so on

so If you have 1 message received and your Multi returns 2 for each message received, 2*1=2 and so on

> If the inputs are different the assertions should also have changed, no?

`Connection` objects contains what is send + rule that responses are prefix with `Response number: `, this way, you can determine bytes and number of messages sent / received.

I'd like to stress that explicit number are fine, but it will mean a lot of repetition. If I want to assert something works, it's easier to assert one method than assert logic in every unit method. Anyway, I am ready to change it if you want.

Reply:
Looks like a good case for parameterized tests.
You have a test and change the inputs and expectations... Can scale to a very large nr of tests.

Reply:
true, I'll think about it; it probably cannot be done everywhere, but there, I can make it explicit

Reply:
1. response multiplier is gone
2. great deal of tests is parametrized

tomorrow I'll look if other tests can be smartly parametrized, not sure if it is really needed but I'd like to address it to your requirements

Reply:
Much better to understand what's going on. Thanks!